PR-Put and Post requests for FioriDAST (#79)
SAP/odfuzz: doc/http_methods.rst

======================
HTTP Methods in ODFUZZ
======================

> Brief Overview of HTTP verb implementation

Reviewer: +1

------------------------------------------

.. csv-table:: HTTP verb implementation
   :header: "", "GET", "DELETE", "PUT", "POST" 
   :widths: 30, 15, 15, 15, 15

   "$inlinecount", "Yes", "No", "No", "No"
   "$search", "Yes", "No", "No", "No"
   "$top", "Yes", "No", "No", "No"
   "$skip", "Yes", "No", "No", "No"
   "$orderby", "Yes", "No", "No", "No"
   "$expand", "Yes", "No", "No", "No"
   "$filter", "Yes", "No", "No", "No"
   "$format", "Yes", "No", "No", "No"
   "body value", "Yes", "Yes", "Yes", "Yes"
   "populated body", "No", "No", "Yes", "Yes"
   "Addressing Single entities", "Yes", "Yes", "Yes", "No"
   "Addressing Multiple entities", "Yes", "Yes", "No", "Yes"
   "Synced URI and body values", "No", "No", "Yes", "No"

**"body value"** refers to the fact that a value for the body is returned by *generate()* function. 
**"populated body"** refers to the fact that the value contains data, and isnt an empty JSON parsed into a string.

Introduction
------------

> Odfuzz through the DirectBuilder class offers developers an interface to integrate its query creation functionalities into their project.

Reviewer 2: ```suggestion
The class DirectBuilder provides an interface for developers to integrate the query generation functionalities into their projects.
```

> As of ver. **0.14a2** the supported list of HTTP methods in DirectBuilder are **"GET", "DELETE", "PUT" and "POST"**. 

Reviewer: You can delete the part "As of ver. **0.14a2** " since this information is in Git history anyway and the line "the supported list of HTTP methods in DirectBuilder are..." is enough :)

Reviewer: ok, will do it myself :)

.. code-block:: python

    import logging
    from pathlib import Path
    from odfuzz.restrictions import RestrictionsGroup
    from odfuzz.entities import DirectBuilder
    from odfuzz.fuzzer import SingleQueryable

    def DirectBuilderExample():
        path_to_metadata = Path(__file__).parent.joinpath("metadata.xml")
        metadata_file_contents = path_to_metadata.read_bytes()
        restrictions = RestrictionsGroup(None)
        #DirectBuilder needs to be called with HTTP method, which in this case is PUT
        examplebuilder = DirectBuilder(metadata_file_contents, restrictions,"PUT")
        example_entities = example_builder.build()
        queryable_factory = SingleQueryable
        queries_list = []
        queries_list.clear()

        #this function iterates through the SingleQueryable object returned
        for queryable in example_entities.all():
            entityset_urls_count = len(queryable.entity_set.entity_type.proprties())
            for _ in range(entityset_urls_count):
                q = queryable_factory(queryable, logger, 1)
                queries,body = q.generate()

> As shown above, DirectBuilder needs to called with the metadata file and restrictions file(if there is any, or else None) along with the HTTP method as a String.

Reviewer 2: ```suggestion
As shown in the snippet above, DirectBuilder is initialized with the metadata file, the restrictions, and the HTTP method defined as a string.
```

.. code-block:: python

 queries,body = q.generate() 

> This line returns the 2 parts of the request: the URI query string, and the body as a JSON string. For **GET** and **DELETE**, body is return as an empty JSON i.e. "{}"

Reviewer 2: ```suggestion
This line returns 2 parts of the request: a URI query string, and a body in the JSON format. For **GET** and **DELETE**, the body is equal to an empty JSON string, i.e. "{}".
```

PUT Queries
-----------

For generating PUT queries, DirectBuilder needs to be called with **method = "PUT"**

.. code-block:: python

    examplebuilder = DirectBuilder(metadata_file_contents, restrictions,"PUT")

On calling *SingleQueryable.generate()*, this would return a fuzzed URI string and a JSON string containing all the fuzzed properties. All the keys appear in the URI and are synchronized with the body. 

::
     
     Categories(CategoryID=1714953551,CategoryName='%21%C2%9Dla%C3%92l%24',Description='hz%60%C3%8F%C3%8F%7B%C3%AAi%2Bk%C3%81%C2%A4%C3%96xc%C5%93%C2%A85k%C3%93%2A%C3%B5%C2%BBrLD%2A%E2%80%A1',Picture=binary'ac9916669fAeb2')?sap-client=500   
     
     {"CategoryID": "1714953551", "CategoryName": "%21%C2%9Dla%C3%92l%24", "Description": "hz%60%C3%8F%C3%8F%7B%C3%AAi%2Bk%C3%81%C2%A4%C3%96xc%C5%93%C2%A85k%C3%93%2A%C3%B5%C2%BBrLD%2A%E2%80%A1", "Picture": "YmluYXJ5J2FjOTkxNjY2OWZBZWIyJw=="}

==================
Code Documentation
==================

> Adding method parameter in DirectBuilder

Reviewer 2: ```suggestion
The **method** parameter in DirectBuilder
```

----------------------------------------

DirectBuilder now has an additional parameter called **method**. This accepts the users choice of HTTP method to get the fuzzed requests. The DirectBuilder *init* checks for validity and calls the *Config.fuzzer* setter to set the value in the Config object.

.. code-block:: python

    def __init__(self, metadata, restrictions,method):
        if method not in ["GET","DELETE","PUT","POST"]:
            raise ValueError("The http method value \'{}\' is invalid\nUse either GET, DELETE, PUT or POST".format(method))
        self._queryable = QueryableEntities()
        self._metadata_string = metadata
        self._restrictions = restrictions
        Config.init()
        Config.fuzzer.http_method_enabled = method

> In config.py, this would be used to build the Config object, which would be looked up during query construction and check which HTTP method is set.

Reviewer: This block should be on the top of the document, after Introduction section, since after enumerating all the possible methods and how to call them, servers no purpose. On the contrary, before diving to differences between HTTP methods, this is introduction to the class usage and it is generic for all http methods.

Truncating Query Options
------------------------

> For Odata queries other than **GET**, query options need to be truncated. To implement this, in fuzzer.py *Query.build_string()* would first check if *Config.fuzzer.http_method_enabled == "GET"* before generating the options and appending them. 

Reviewer 2: ```suggestion
Except for **GET** requests, query options need to be truncated in OData. To implement this, *Query.build_string()* first checks if *Config.fuzzer.http_method_enabled == "GET"* before generating the options and appending them.
```

I am not sure whether such implementation details should be mentioned in the documentation or not.

Reviewer: I am ok with that. I would like to split in the future the docs into two distinct sections... for ODfuzz users.. and for odfuzz developers. This entire .rst file is ment to be part of the second. 

It is true that highlighted block could be obsolete soon.. and maybe should be just explaining things generally and point to code via permalink.. so if changed it will be understandable possible discrepancy between docs and code. But it is ok to polish the docs in separate PR if needed, so not to block main purpose of the proposed change.

The list of options are illustrated in the table above.

Generating a Body
-----------------

For PUT and POST queries, a new element for the queries need to be generated i.e. the **body** of the request. 
The first change is returning a tuple of *query, body* instead of just the *query* from *SingleQueryable.generate()*.
A new function *generate_body()* is added which fetches the proprties from the metadata and calls the generator on each of them, and appends them is a dictionary.

> The dictionary is jsonified before being returned as the body. This process is skipped for **GET** and **DELETE**,and they return an empty jsonified string instead, to be compatible with the tuple returned. The changes made to the generators are described further below.

Reviewer: "the dictionary is jsoified befor being returned as the body" - does this mean that you expects ONLY JSON as possible format? It is possible that your changes were tested only against odata services that produces and consumes just the JSON option. 

But Odata protocol - and odfuzz is trying to cover that- is having both XML/Atom representation and JSON representation. https://www.odata.org/documentation/odata-version-2-0/atom-format/

the CLI configuration options
```
export ODFUZZ_DATA_FORMAT=json
export ODFUZZ_DATA_FORMAT=xml 
```

Check the usage of: 
https://github.com/SAP/odfuzz/blob/bf08492ce8914a56f27a7f7bd3d0e5352064e722/odfuzz/config.py#L47:

@AlapanGhosh Please confirm my suspicion. If yes, I have no problem merging as it currently is, marking another alpha version, but it is another task to cover in some next PR.

Reviewer 3: @phanak-sap yes, currently the body is generated in only JSON. Wont be problem to parse into XML either, and the config already exists, but a setter needs to be added.  Best to have a separate PR.

Reviewer: @AlapanGhosh I agree that the scope of the PR is getting bigger and bigger by almost each comment here :)
I have tracked this as another task for you under the https://github.com/SAP/odfuzz/issues/7

Differentiating between PUT and POST
------------------------------------

> PUT is idempotent and address single entities, whereas POST isnt idempotent and addresses multiple entities. Therefore changes are made in *DirectBuilder._append_queryable()* so that PUT avoids generating multiple entities and POST avoids generating single entities, during query generation.

Reviewer: I am confused by the idempotency, which difference is marked, but not addressed in any way.

What exactly do you mean by difference between PUT and POST idempotency? Why is then explained only difference in Single and Multiple entity usage, but not the idempotency difference?

I mean, yes, it is the property of PUT/POST in HTTP protocol - but it is a generic property and it basically about how the target service should behave. Or does it affect the ODfuzz code in any way? I don't see it in the PR. 

If yes, pls explain futher. If no, perhaps would be better to ditch the idempotency from the docs and focus only on the differences between the single/multiple entities.

Reviewer 3: Yes, it refers to how the service should ideally behave, which is reflected in the choice between single/multiple entities while generating payloads in ODfuzz. I'll trim it down to precisely mention single/multiple entities

.. code-block:: python

    def _append_queryable(self, query_group_data):
        # TODO REFACTOR DRY this method is direct copypaste from DispatchedBuilder just to have a prototype for integration. Intentionally no abstract class at the moment.
        if Config.fuzzer.http_method_enabled != "POST":
            self._append_corresponding_queryable(QueryGroupSingle(query_group_data))
        if Config.fuzzer.http_method_enabled != "PUT":
            self._append_corresponding_queryable(QueryGroupMultiple(query_group_data))
            self._append_associated_queryables(query_group_data)

Alternative EDM Generators for Body
-----------------------------------

> Some EDM data types have different representation format in the body than in the URI. So the generators needed to adapt for the body implementation. The *generate()* function in the generator classes now have an additional parameter **format** which is provided the value "body". The default generation is done by **generate(format="uri")**. Following is an example of calling generator on a property for body format.

Reviewer 2: It would be great to provide some examples of how these different representations look like. I have no idea what representations you are referring to.

Reviewer: yeah, actually unit tests are great fist examples of how to use it and what are the real differences .)

Reviewer: @AlapanGhosh Please do not forget this in the next PR.
